common/src/main/java/us/ilite/common/lib/control/PIDController.java

```diff
@@ -0,0 +1,320 @@
+     * setpoint.
+     *
+     * @param continuous
+     *            Set to true turns on continuous, false turns off continuous
+     */
+    public void setContinuous(boolean continuous) {
+        m_continuous = continuous;
+    }
+
+    public void setDeadband(double deadband) {
+        m_deadband = deadband;
+    }
+
+    /**
+     * Set the PID controller to consider the input to be continuous, Rather then using the max and min in as
+     * constraints, it considers them to be the same point and automatically calculates the shortest route to the
+     * setpoint.
+     */
+    public void setContinuous() {
+        this.setContinuous(true);
+    }
+
+    /**
+     * Sets the maximum and minimum values expected from the input.
+     *
+     * @param minimumInput
+     *            the minimum value expected from the input
+     * @param maximumInput
+     *            the maximum value expected from the output
+     */
+    public void setInputRange(double minimumInput, double maximumInput) {
+        if (minimumInput > maximumInput) {
+            System.out.println("Lower bound is greater than upper bound");
+        }
+        m_minimumInput = minimumInput;
+        m_maximumInput = maximumInput;
+        setSetpoint(m_setpoint);
+    }
+
+    /**
+     * Sets the minimum and maximum values to write.
+     *
+     * @param minimumOutput
+     *            the minimum value to write to the output
+     * @param maximumOutput
+     *            the maximum value to write to the output
+     */
+    public void setOutputRange(double minimumOutput, double maximumOutput) {
+        if (minimumOutput > maximumOutput) {
+            System.out.println("Lower bound is greater than upper bound");
+        }
+        m_minimumOutput = minimumOutput;
+        m_maximumOutput = maximumOutput;
+    }
+
+    /**
+     * Set the setpoint for the PID controller
+     *
+     * @param setpoint
+     *            the desired setpoint
+     */
+    public void setSetpoint(double setpoint) {
+        m_setpoint = Math.min( m_maximumInput, setpoint );
+        m_setpoint = Math.max( m_minimumInput, setpoint );
+    }
+
+    /**
+     * Returns the current setpoint of the PID controller
+     *
+     * @return the current setpoint
+     */
+    public double getSetpoint() {
+        return m_setpoint;
+    }
+
+    /**
+     * Returns the current difference of the input from the setpoint
+     *
+     * @return the current error
+     */
+    public double getError() {
+        return m_error;
+    }
+
+    /**
+     * Return true if the error is within the tolerance
+     *
+     * @return true if the error is less than the tolerance
+     */
+    public boolean onTarget(double tolerance) {
+        return m_last_input != Double.NaN && Math.abs(m_last_input - m_setpoint) < tolerance;
+    }
+
+    /**
+     * Reset all internal terms.
+     */
+    public void reset() {
+        m_last_input = Double.NaN;
```

Comment:
Are we properly handling NaN? If any calculation acts on a Nan, the result will be NaN.

Reply:
As far as I know yes, calculations are only performed once m_last_input is set